main.go

```diff
@@ -115,21 +115,23 @@ func main() {
 	// Acquire environmental defaults and initialize operator defaults with them
 	placementv1.SetupDefaults()
 
+	checker := healthz.Ping
 	// Setup webhooks if requested
 	if strings.ToLower(os.Getenv("ENABLE_WEBHOOKS")) != "false" {
 		if err = (&placementv1.PlacementAPI{}).SetupWebhookWithManager(mgr); err != nil {
 			setupLog.Error(err, "unable to create webhook", "webhook", "PlacementAPI")
 			os.Exit(1)
 		}
+		checker = mgr.GetWebhookServer().StartedChecker()
 	}
 
 	//+kubebuilder:scaffold:builder
 
-	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+	if err := mgr.AddHealthzCheck("healthz", checker); err != nil {
 		setupLog.Error(err, "unable to set up health check")
 		os.Exit(1)
 	}
-	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+	if err := mgr.AddReadyzCheck("readyz", checker); err != nil {
 		setupLog.Error(err, "unable to set up ready check")
 		os.Exit(1)
 	}
```

Comment:
Is this change fixes an actual bug in the system? If so then I can be convinced to land this before the dev preview.

Reply:
This fixes the api failures caused by non-ready webhook, which we initially found in edpm jobs.

example:

https://logserver.rdoproject.org/15/215/520a66159a28fba5a4d21271d217e493572436c2/github-check/edpm-ansible-crc-podified-edpm-baremetal/28c43f7/ci-framework-data/logs/ci_make_005_run_edpm_baremetal.log
```
Internal error occurred: failed calling webhook "mnetconfig.kb.io": failed to call webhook: Post "https://infra-operator-controller-manager-service.openstack-operators.svc:443/mutate-network-openstack-org-v1beta1-netconfig?timeout=10s": dial tcp 10.217.0.93:9443: connect: connection refused
```

The above example error is triggered by infra-operator but this problem can occur in all operators now.

Reply:
Thanks for the pointers.
So from the log:
```
++ oc get -n openstack-operators csv/infra-operator.v0.0.1 -o 'jsonpath={.status.phase}'
+ PHASE=Succeeded
+ echo Succeeded
+ '[' Succeeded '!=' Succeeded ']'
+ exit 0
+ rc=0
+ popd
+ exit 0
+ '[' -z /home/zuul/ci-framework-data/artifacts/manifests/openstack/infra/cr ']'
+ NEXT_WAIT_TIME=0
+ '[' 0 -eq 15 ']'
+ oc kustomize /home/zuul/ci-framework-data/artifacts/manifests/openstack/infra/cr
+ oc apply -f -
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "mnetconfig.kb.io": failed to call webhook: Post "https://infra-operator-controller-manager-service.openstack-operators.svc:443/mutate-network-openstack-org-v1beta1-netconfig?timeout=10s": dial tcp 10.217.0.93:9443: connect: connection refused
```
The script waits for the .status.phase to become "Succeeded" before it tries to use the operator. Will the proposed change prevents the operator to reach "Succeeded" state if the webhooks are not started yet? What will be the phase while the operator waits for the webhooks?

Reply:
That phase become "Succeeded" after the controller-manager deployment becomes ready and the controller-manage deployment becomes ready when the controller-manager pod is launched and it responds to the healthcheck using the `readyz` healthcheck.

https://github.com/openstack-k8s-operators/placement-operator/blob/main/config/manager/manager.yaml#L40-L51

Current problem is that these healthchecks (both readyz for readiness and healthz for liveness) are doing null check so the pod/deployment becomes ready once it launches the healthcheck endpoints and these is no mechanism to check availability/readiness of webhook.

With this change, readyz healthcheck returns success only after webhook severer becomes ready, and deployment readiness and csv phase is updated only after webhook server is started and becomes ready.

Reply:
Thanks. I'll bring this to our sync call tomorrow to get attention from the team and hopefully agreement with merging this early as a bug fix.